core/src/main/java/io/questdb/cairo/map/SpillableHeapMemory.java

```diff
@@ -0,0 +1,402 @@
+ * The extend() behavior follows the underlying MemoryCARWImpl and MemoryCMARWImpl, which has different behavior
+ * in some edge cases.
+ */
+public class SpillableHeapMemory implements MemoryCARW, Mutable {
+    private static final Log LOG = LogFactory.getLog(SpillableHeapMemory.class);
+    // flag
+    private final long diskSpillThreshold;
+    private final long extendSegmentMsb;
+    private final FilesFacade ff = new FilesFacadeImpl();
+    private final int maxPages;
+    private final long maxSize;
+    private final int nativeMemoryTag;
+    // for non-mmap
+    private final long pageSize;
+    // general
+    private final int spillMemoryTag;
+    // for mmap
+    private final Path spillPath;
+    private boolean isSpilled;
+    private MemoryCARW mem;
+    private long opts = CairoConfiguration.O_NONE;
+
+    public SpillableHeapMemory(long pageSize,
+                               long diskSpillThreshold,
+                               long extendSegmentSize,
+                               long maxSize,
+                               Path spillFilePath) {
+        this(pageSize,
+                diskSpillThreshold,
+                extendSegmentSize,
+                maxSize,
+                spillFilePath,
+                MemoryTag.NATIVE_DEFAULT,
+                MemoryTag.MMAP_DEFAULT);
+    }
+
+    SpillableHeapMemory(long pageSize,
+                        long diskSpillThreshold,
+                        long extendSegmentSize,
+                        long maxSize,
+                        Path spillFilePath,
+                        int nativeMemoryTag,
+                        int spillMemoryTag) {
+        this.pageSize = Numbers.ceilPow2(pageSize);
+        extendSegmentSize = Numbers.ceilPow2(extendSegmentSize);
+        this.extendSegmentMsb = Numbers.msb(extendSegmentSize);
+        this.maxSize = Numbers.ceilPow2(maxSize);
+        this.diskSpillThreshold = Numbers.ceilPow2(diskSpillThreshold);
+
+        assert this.maxSize > this.pageSize :
+                String.format("maxSize (%s) should be larger than pageSize (%s)",
+                        this.maxSize, this.pageSize);
+        assert diskSpillThreshold > pageSize :
+                "diskSpillThreshold should be larger than pageSize";
+        if (maySpill()) {
+            // if it is configured not to spill, extendSegmentMsb is irrelevant
+            assert this.extendSegmentMsb > Numbers.msb(this.pageSize) :
+                    "extendSegmentMsb should be larger than pageSizeMsb";
+            assert maxSize > (1L << this.extendSegmentMsb) :
+                    "maxSize should be larger than extendSegmentSize";
+        }
+
+        this.maxPages = getMaxPages(this.maxSize, this.pageSize);
+
+        this.nativeMemoryTag = nativeMemoryTag;
+        this.spillMemoryTag = spillMemoryTag;
+        mem = initMem();
+        spillPath = spillFilePath;
+
+        // pre-check spillPath can be used
```

Comment:
Well, since ff.touch(spillPath) can throw exception, I thought pretest during initialization might be helpful rather than let it unnoticeable until by the time its started spilling and fail. wdyt?

Reply:
No, most methods in `FilesFacade` don't throw. Instead, the result has to be checked. But my point is that this check is expensive considering that in most cases we don't expect disk spill to happen. Also, even if this check is successful, there is a chance that later file operations will fail.